polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
+      SmallVector<AffineExpr, 4> symbols;
+      for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumSymbols();
+           ++idx)
+        symbols.push_back(getAffineSymbolExpr(
+            idx + loop.getLowerBoundsMap().getNumSymbols(), loop.getContext()));
+
+      SmallVector<AffineExpr, 4> dims;
+      for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumDims(); ++idx)
+        dims.push_back(getAffineDimExpr(
+            idx + loop.getLowerBoundsMap().getNumDims(), loop.getContext()));
+
+      ub = ub.replaceDimsAndSymbols(dims, symbols);
+
+      for (AffineExpr lb : loop.getLowerBoundMap(step.index()).getResults()) {
+        // Bound is whether this expr >= 0, which since we want ub > lb, we
+        // rewrite as follows.
+        exprs.push_back(ub - lb - step.value());
+        eqflags.push_back(false);
       }
     }
-
-    SmallVector<Value> values;
-    auto lb_ops = looplike.getLowerBoundsOperands();
-    auto ub_ops = looplike.getUpperBoundsOperands();
-    for (unsigned idx = 0; idx < looplike.getLowerBoundsMap().getNumDims();
-         ++idx) {
-      values.push_back(lb_ops[idx]);
-    }
-    for (unsigned idx = 0; idx < looplike.getUpperBoundsMap().getNumDims();
-         ++idx) {
-      values.push_back(ub_ops[idx]);
-    }
-    for (unsigned idx = 0; idx < looplike.getLowerBoundsMap().getNumSymbols();
-         ++idx) {
-      values.push_back(lb_ops[idx + looplike.getLowerBoundsMap().getNumDims()]);
-    }
-    for (unsigned idx = 0; idx < looplike.getUpperBoundsMap().getNumSymbols();
-         ++idx) {
-      values.push_back(ub_ops[idx + looplike.getUpperBoundsMap().getNumDims()]);
-    }
-
-    auto iset = IntegerSet::get(
-        /*dim*/ looplike.getLowerBoundsMap().getNumDims() +
-            looplike.getUpperBoundsMap().getNumDims(),
-        /*symbols*/ looplike.getLowerBoundsMap().getNumSymbols() +
-            looplike.getUpperBoundsMap().getNumSymbols(),
-        exprs, eqflags);
-    auto ifOp = b.create<AffineIfOp>(looplike.getLoc(), TypeRange(), iset,
-                                     values, /*else*/ false);
-    looplike->moveBefore(ifOp.getThenBlock()->getTerminator());
   }
-  for (auto op : opsToMove)
-    looplike.moveOutOfLoop(op);
-  LLVM_DEBUG(looplike.print(llvm::dbgs() << "\n\nModified loop:\n"));
+
+  SmallVector<Value> values;
+  OperandRange lb_ops = loop.getLowerBoundsOperands(),
+               ub_ops = loop.getUpperBoundsOperands();
+  for (unsigned idx = 0; idx < loop.getLowerBoundsMap().getNumDims(); ++idx)
+    values.push_back(lb_ops[idx]);
+  for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumDims(); ++idx)
+    values.push_back(ub_ops[idx]);
+  for (unsigned idx = 0; idx < loop.getLowerBoundsMap().getNumSymbols(); ++idx)
+    values.push_back(lb_ops[idx + loop.getLowerBoundsMap().getNumDims()]);
+  for (unsigned idx = 0; idx < loop.getUpperBoundsMap().getNumSymbols(); ++idx)
+    values.push_back(ub_ops[idx + loop.getUpperBoundsMap().getNumDims()]);
```

Comment:
Can we use something like `std::copy` here instead?

Reply:
Ok, done. Please check the new code.